lib/jsonrpc2.ex

```diff
@@ -152,11 +152,10 @@ defmodule JSONRPC2 do
           | integer
           | String.t()
           | [json]
-          | %{optional(String.t()) => json}
-          | %{optional(atom()) => json}
+          | %{optional(String.t() | atom()) => json}
```

Comment:
Shouldn't this always be String, since we are decoding? There should not be any atoms in the result, unless your decoder is converting strings to atoms, which can be dangerous.

Reply:
We are using `jason` as the serializer which comes as default and is accepting nested maps with atom keys to encode. Our use case scenario which dialyzer complains is actually to encode, using the `JSONRPC2.Clients.HTTP.call` function.

Maybe another alternative I see here is to create another type of params just for JSON serialization, or just change the `typedoc` to make it clear that it is for both encoding/decoding a JSON object, what are your thoughts?

Reply:
Actually, the `Jason.encode` accepts any `term`, and I think the params type should be flexible to also accept any term from any serializer:

```
  @spec encode(term, [encode_opt]) ::
          {:ok, String.t()} | {:error, EncodeError.t() | Exception.t()}
```

What do you think if we just change the type params to the `term` type?

Reply:
Like this:

```elixir
  @typedoc "A decoded JSON object."
  @type json ::
          nil
          | true
          | false
          | float
          | integer
          | String.t()
          | [json]
          | %{optional(String.t()) => json}

  @typedoc "A JSON-RPC 2.0 params value."
  @type params :: term()
```

Reply:
The only concern I have is that valid JSON-RPC params are either a list, or a map. How about this:

```
  @typedoc "A decoded JSON object."
  @type json ::
          nil
          | true
          | false
          | float
          | integer
          | String.t()
          | [json]
          | %{optional(String.t()) => json}

  @typedoc "A JSON-RPC 2.0 params value."
  @type params :: [term] | %{optional(term) => term}
```

Reply:
lgtm!

Reply:
Looks better, but I think this makes more sense, since maps can only be string or atom:

```elixir
  @type params :: [term] | %{optional(String.t() | atom()) => term}
```